DRIVERS-3298 Improve stop-orchestration behavior

blink1073 opened the pull request:
[DRIVERS-3298](https://jira.mongodb.org/browse/DRIVERS-3298)

Replaces #690.

cc @baileympearson

eramongodb commented on .evergreen/orchestration/drivers_orchestration.py:
```diff
@@ -635,36 +635,83 @@ def start(opts):
+
+
+def shutdown_proc(proc: psutil.Process) -> None:
+    proc.terminate()
+    if proc.wait(1) is None:
+        try:
+            proc.kill()
+        except psutil.NoSuchProcess:
+            pass
```
```suggestion
	try:
		proc.terminate()
		try:
		    proc.wait(10) # Wait up to 10 seconds.
		except psutil.TimeoutExpired:
		    proc.kill()
	except psutil.NoSuchProcess:
		pass
```

A 1 second timeout seems really short. This may throw psutil.TimeoutExpired on failure to terminate the process within 1 second. Docs also state:

> On UNIX: If PID is not a child of os.getpid (current process) just wait until the process disappears and return None. If PID does not exist return None immediately.

which doesn't seem consistent with the current check.

Should this instead be implemented as suggested?

blink1073 commented on .evergreen/orchestration/pyproject.toml:
```diff
@@ -17,7 +17,10 @@ classifiers = [
+dependencies = [
+  "drivers-evergreen-tools @ {root:parent:uri}",
+  "mongo-orchestration @ https://github.com/mongodb/mongo-orchestration/archive/master.tar.gz",
+  "psutil>=6.0"]
```
I wanted a minimum for dependabot, so I chose one about a year old as a starting point.

eramongodb replied:
Unless there are platform constraints (e.g. cannot obtain the dependency on some EVG hosts), suggest bumping to the latest version 7.1.0, since there appear to be bugfixes and nice-to-have enhancements such as:

- 6.1.0: drastically speedup `process_iter()`
- 6.1.1: fix race condition in case `/proc/PID/stat` does not exist, but `/proc/PID does`, resulting in `FileNotFoundError`.
- 7.1.0: [macOS], [CRITICAL]: fixed different places in C code which can trigger a segfault.

blink1073 commented on .evergreen/orchestration/drivers_orchestration.py:
```diff
@@ -635,36 +635,83 @@ def start(opts):
+    # Next look for runner docker images.
+    if docker:
+        cmd = "docker ps --format '{{.Image}}\t{{.ID}}'"
+        response = subprocess.check_output(cmd, shell=True, encoding="utf-8").strip()
```
Ah, I needed to split the command with shlex, fixed

eramongodb commented on .evergreen/orchestration/drivers_orchestration.py:
```diff
@@ -638,14 +638,14 @@ def start(opts):
-		    proc.kill()
-	except psutil.NoSuchProcess:
-		pass
+    try:
+        proc.terminate()
+        try:
+            proc.wait(10)  # Wait up to 10 seconds.
+        except psutil.TimeoutExpired:
+            proc.kill()
+    except psutil.NoSuchProcess:
+        pass
```
Oof, I'll have to remember to avoid this when using GitHub's new PR editor.

blink1073 commented:
Okay, I tested again.  I had to escape the quotes in the `docker ps` command to get it working with `shlex.split`.  I also found that if I started `mongo-orchestration start` manually in another terminal it had a different cmdline signature, so I added handling for that.

blink1073 commented on .evergreen/orchestration/drivers_orchestration.py:
```diff
@@ -635,36 +637,86 @@ def start(opts):
+
+    # Next look for running docker images.
+    if docker:
+        cmd = f"{docker} ps --format '" "{{.Image}}\t{{.ID}} " "'"
```
I confirmed this still works

blink1073 commented:
I had to account for some platform-specific errors/concerns, it should be good to go now.

eramongodb commented on .evergreen/orchestration/drivers_orchestration.py:
```diff
@@ -198,6 +198,8 @@ def get_docker_cmd():
     docker = shutil.which("podman") or shutil.which("docker")
     if not docker:
         return None
+    if os.name == "nt":
+        docker = docker.replace(os.sep, "/")
```
```suggestion
    docker = PureWindowsPath(docker).as_posix()
```
I think this should reliably convert any Windows `'\\'` to `'/'` regardless of the OS. Needs `from pathlib import PureWindowsPath`.

eramongodb reviewed (approved):
Minor suggestion; otherwise, LGTM.
